buffer/src/main/java/io/netty/buffer/api/Buffer.java

```diff
@@ -0,0 +1,682 @@
+     * <p>
+     * See the <a href="#split">Splitting buffers</a> section for details.
+     *
+     * @return A new buffer with independent and exclusive ownership over the read and readable bytes from this buffer.
+     */
+    default Buffer split() {
+        return split(writerOffset());
+    }
+
+    /**
+     * Split the buffer into two, at the given {@code splitOffset}.
+     * <p>
+     * The region of this buffer that precede the {@code splitOffset}, will be captured and returned in a new
+     * buffer, that will hold its own ownership of that region. This allows the returned buffer to be independently
+     * {@linkplain #send() sent} to other threads.
+     * <p>
+     * The returned buffer will adopt the {@link #readerOffset()} and {@link #writerOffset()} of this buffer,
+     * but truncated to fit within the capacity dictated by the {@code splitOffset}.
+     * <p>
+     * The memory region in the returned buffer will become inaccessible through this buffer. If the
+     * {@link #readerOffset()} or {@link #writerOffset()} of this buffer lie prior to the {@code splitOffset},
+     * then those offsets will be moved forward, so they land on offset 0 after the split.
+     * <p>
+     * Effectively, the following transformation takes place:
+     * <pre>{@code
+     *         This buffer:
+     *          +--------------------------------+
+     *         0|               |splitOffset     |cap
+     *          +---------------+----------------+
+     *         /               / \               \
+     *        /               /   \               \
+     *       /               /     \               \
+     *      /               /       \               \
+     *     /               /         \               \
+     *    +---------------+           +---------------+
+     *    |               |cap        |               |cap
+     *    +---------------+           +---------------+
+     *    Returned buffer.            This buffer.
+     * }</pre>
+     * When the buffers are in this state, both of the split parts retain an atomic reference count on the
+     * underlying memory. This means that shared underlying memory will not be deallocated or returned to a pool, until
+     * all the split parts have been closed.
+     * <p>
+     * Composite buffers have it a little easier, in that at most only one of the constituent buffers will actually be
+     * split. If the split point lands perfectly between two constituent buffers, then a composite buffer can
+     * simply split its internal array in two.
+     * <p>
+     * Split buffers support all operations that normal buffers do, including {@link #ensureWritable(int)}.
+     * <p>
+     * See the <a href="#split">Splitting buffers</a> section for details.
+     *
+     * @return A new buffer with independent and exclusive ownership over the bytes from the beginning to the given
+     * offset of this buffer.
+     */
+    Buffer split(int splitOffset);
```

Comment:
The name `split()` makes me feel like it will return two `Buffer`s, like Scala's [`splitAt()`](https://www.geeksforgeeks.org/scala-list-splitat-method-with-example/). Could we have a better name?

I also wonder if it'd be useful to have a method that actually returns two `Buffer`s, when a user wants to read and write each part.

Reply:
That could be a Scala-ism. Multiple return types are not customary in Java. Rust has a [split_off](https://doc.rust-lang.org/std/vec/struct.Vec.html#method.split_off) method which is similar to this one, but returning the end-region and leaving the original vector with the start, while this split method does the opposite, which seems more sensible for our use cases.

This method was previously called `bifurcate` because I though it would be rarely used, but now I believe it will be used quite frequently, and then `split` is a better, more straight forward name.

Reply:
Fair enough. At second thought, it sounds reasonable. :smile: